### PR TITLE
add bag.fold operator

### PR DESCRIPTION
**Pull request by mudathirmahgoub**

### CODE REVIEW

File: src/theory/bags/theory_bags.cpp [hunk 1]
```diff
@@ -152,6 +156,114 @@ TrustNode TheoryBags::expandCardOperator(TNode n,
...
+};
+typedef expr::Attribute<IndexVarAttributeId, Node> IndexVarAttribute;
+
+TrustNode TheoryBags::expandFoldOperator(TNode node,
+                                         std::vector<SkolemLemma>& lems)
+{
+  Assert(node.getKind() == BAG_FOLD);
+  if (d_env.getLogicInfo().isHigherOrder())
+  {
+    NodeManager* nm = NodeManager::currentNM();
```
[the remark this answers is not in the document]
Comment (mudathirmahgoub) [reply src/theory/bags/theory_bags.cpp]: Added file `bag_reduction.h,cpp` for class `BagReduction`. For now it only reduces the fold operator. Other operators (bag.choose, bag.card, bag.map) would be refactored in future PR.

File: src/theory/bags/bag_reduction.h [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+   *   n: Int is the cardinality of bag A
+   *   uf:Int -> T1 is an uninterpreted function that represents elements of A
+   *   combine: Int -> T2 is an uninterpreted function
+   *   unionDisjoint: Int -> (Bag T1) is an uninterpreted function
+   */
+  Node reduceFoldOperator(Node node, std::vector<Node>& asserts);
+
+ private:
+  /** The associated environment. */
+  Env& d_env;
```
Comment (ajreynol) [inline src/theory/bags/bag_reduction.h:71]: As a style point here, you should make this class inherit from EnvObj, and not have an explicit reference d_env.

**Review by ajreynol — CHANGES_REQUESTED** [review]
One minor request.

File: src/theory/bags/bag_reduction.h (hunk 2 above)
Comment (mudathirmahgoub) [reply src/theory/bags/bag_reduction.h]: Done.

**Review by ajreynol — APPROVED** [review]

**Pull request merged**